### ISSUE TITLE
true REST semantics

### ISSUE THREAD

**Issue by mdutoo** [cgourdin/MartServer#51]
See https://github.com/cgourdin/MartServer/issues/47#issuecomment-264481977

- POST /category should create in the category, but for now it fails. POST /(category or full path/)id could be accepted in a lenient way, maybe with PATCH semantics to make it useful differently from PUT.
- PUT /(category or full path/)id correctly creates (though I personally don't like it) or update. However it should return what was sent, for now it's empty instead. PUT /(category or full path/) should not be accepted.

REST semantics exist for a reason: it makes HTTP features such as client-side cache work, it allows users to implicitly know how REST servers work...

**Comment by cgourdin** [comment]
I'll respect the OCCI specification see page 10 of the http_protocol documentation.
By the way, there is an issue with POST, in MartServer, on resource creation, if you dont have the location, is not implemented and only PUT method is usable.
The usage of POST on a root path is the only way to create a resource without a location. The resource will be in a path like /my_kind/resourceID.
PATCH method should be used to partial update only (and so this operation should be removed from POST).
I'm agree with "POST /" should create the resource in the kind path like /compute/resourceID.
A good article (in French) on the POST / PUT method is described here : http://blog.xebia.fr/2014/03/17/post-vs-put-la-confusion/

In conclusion, MART server must authorize resource creation with POST method only when no location path is defined (like : POST / ).
MART server must use PATCH method (and not POST) to update partially a resource, but be aware that **this is not specified in OCCI specification**.
MART server must use PUT to create a resource with a location path defined and use PUT to totally replace an existing resource.
PUT method must return what was sent.
Multiple PUT on the same resource must be idempotent and return always the same thing (same uuid same attributes, same location etc.).

**Comment by mdutoo** [comment]
Happy to se that you know the spec much better than me :)
Let's not start endless philosophizing about REST anyway. I just wanted to point out that
- doing a PUT elsewhere that on a resource is wrong (ex. PUT /compute is wrong but PUT /compute/id is OK), otherwise PUT idempotency is not respected
- what OCCInterface allows is what erocci allowed, because of spec but also IN ORDER TO HELP USERS which is my second point.
